psutil/_psutil_sunos.c

```diff
@@ -154,6 +156,79 @@ psutil_proc_name_and_args(PyObject *self, PyObject *args) {
     return NULL;
 }
 
+/*
+ * Return process environ block
+ */
+static PyObject *
+psutil_proc_environ(PyObject *self, PyObject *args) {
+    int pid;
+    char path[1000];
+    psinfo_t info;
+    const char *procfs_path;
+    char **env = NULL;
+    ssize_t env_count = -1;
+    char *dm;
+    int i = 0;
+    PyObject *py_retdict = NULL;
+    PyObject *py_envname = NULL;
+    PyObject *py_envval = NULL;
+
+    if (! PyArg_ParseTuple(args, "is", &pid, &procfs_path))
+        goto error;
+
+    sprintf(path, "%s/%i/psinfo", procfs_path, pid);
+    if (! psutil_file_to_struct(path, (void *)&info, sizeof(info)))
+        goto error;
+
+    env = psutil_read_raw_env(info, procfs_path, &env_count);
+    if (! env && env_count != 0)
+        goto error;
+
+    py_retdict = PyDict_New();
+    if (! py_retdict)
+        goto enomem;
+
+    for (i=0; i<env_count; i++) {
+        if (! env[i])
+            break;
+
+        /* Environment corrupted */
+        dm = strchr(env[i], '=');
+        if (! dm)
+            break;
+
+        *dm = '\0';
+
+        py_envname = PyUnicode_DecodeFSDefault(env[i]);
+        if (! py_envname)
+            goto error;
+
+        py_envval = PyUnicode_DecodeFSDefault(dm+1);
+        if (! py_envname)
+            goto error;
+
+        if (PyDict_SetItem(py_retdict, py_envname, py_envval) < 0)
+            goto error;
+
+        Py_DECREF(py_envname);
+        Py_DECREF(py_envval);
+    }
+
+    psutil_free_cstrings_array(env, env_count);
+    return py_retdict;
+
+ enomem:
+    PyErr_NoMemory();
```

Comment:
Because before `malloc` you may already have allocated other resources. Or maybe you didn't, but you change your code later and open a fd or use another `malloc` before this one, or create a Python list or whatever.
The "free" logic should just live in one place (`error:`) so that every time you want to raise an exception you set the error type (`PyErr_NoMemory();` or whatever) and then just do `goto error`.
This way you don't have to reason about what resources you allocated thus far and you don't have to infer whether to use `goto errmem`, `goto oserror` etc.